### ISSUE TITLE
Loss

### ISSUE THREAD

**Issue by djo-koconi** [anicolson/DeepXi#20]
Hi Aaron.

Could you please help me understand the loss and val error?

`
val_error_mbatch = sess.run(net.loss, feed_dict={net.input_ph: mbatch[0],
		net.target_ph: mbatch[1], net.nframes_ph: mbatch[2], net.training_ph: False}) # validation error for each frame in mini-batch.
`
So mini-batch contains: x_MAG-fr (noisy), xi_bar (target xi), L .

I'm unsure at which step is network making a prediction about the SNR.
Or it is just learning how to map the target SNR to a defined value and just apply predicted value in inference?

**Comment by anicolson** [comment]
Thats correct

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by anicolson** [comment]
Do you have access to “Deep learning for minimum mean-square error approaches to speech enhancement”? It is explained there.

Hope this helps.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by djo-koconi** [comment]
Thanks, I revisited it.

Actually, I am experimenting with inserting denoising layer (your Deep Xi for example) inside (before) the ASR (as Deep Speech). 
Eventually, I want to train them together so that ASR would have embedded noise reduction layer.

Just by outputting the SNR we would have to do inference before feeding the training signal to the ASR.


Do you think I can make DeepXi to output the predicted clean mag spectrogram of the input signal and feed it forward during the training?

**Comment by anicolson** [comment]
Do be honest, if you are going to train them together, it would be better just to train Deep Speech on noisy speech. Deep Speech will then learn how to remove the noise by itself. 

What would be interesting is if you trained Deep Speech on the output of Deep Xi. You could compute the MFCC coefficients directly from the enhanced speech magnitude spectrum. (Just be warned that the available Deep Xi uses a window size of 32 ms and a shift of 16 ms, and Deep Speech uses something different, I think a window size of 25 ms and a frameshift of 10 ms. So it would be ideal if these matched).

**Comment by djo-koconi** [comment]
Yes, I am trying to do similar to second one, stack Deep Speech layers after Deep Xi  and train the whole network together. 

I am trying to feed the output of the ResNet into the input of DeepSpeech layers.
Just output of ResNet is shape (?,257), while LSTM expects 3 dimensions vector.

Disabling boolean_mask keeps the 3 dimension output, but the program crashes in counting loss. Loss function will be obviously replaced, but I am curious can Deep Xi work with boolean_mask=False?

**Comment by anicolson** [comment]
Not really, it was designed to train with 2d target. As you have sequence length, you could reshape 2d output to 3d (with padding) and then pass to deep speech. This is not ideal but would be the easiest way to do it.

Hope this helps.

**Comment by djo-koconi** [comment]
Thank you Aaron.

I am planning to stack the layers of Deep Xi and Deep speech, just without applying gain the layers may not lead to any noise reduction or accuracy increase.

How couId I estimate the clean signal during the training (of Deep Xi) and feed it forward to DeepSpeech for further training?

**Comment by anicolson** [comment]
It would be better to first train Deep Xi, and then train Deep Speech.

**Comment by djo-koconi** [comment]
I did training with espnet and Deep Xi output, without any benefit.

If you wouldn't mind one more clarification.

So it is possible to estimate enhanced spectrogram during training from mbatch ? 
First dimension is batch of noisy MS, second is the predicted SNR, so by multiplying with a gain would give us the enhanced MS at the training stage, right?

**Comment by anicolson** [comment]
Yea

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by djo-koconi** [comment]
Tnx, did it :)

**Comment by 20050710212** [comment]
> Do be honest, if you are going to train them together, it would be better just to train Deep Speech on noisy speech. Deep Speech will then learn how to remove the noise by itself.
> 
> What would be interesting is if you trained Deep Speech on the output of Deep Xi. You could compute the MFCC coefficients directly from the enhanced speech magnitude spectrum. (Just be warned that the available Deep Xi uses a window size of 32 ms and a shift of 16 ms, and Deep Speech uses something different, I think a window size of 25 ms and a frameshift of 10 ms. So it would be ideal if these matched).

Hi Aaron, I want to train the DeepXi in MFCC features instead of magnitude of each bin. Then, feed the enhanced MFCC features directly to ASR. Since the MFCC features is much less, the network parameters can be reduced. Do you have any suggestions? Thank you.
